src/Functions/naiveBayesClassifier.h

```diff
@@ -0,0 +1,388 @@
+            while (start < text.size() && text[start] == ' ')
+                ++start;
+            if (start == text.size())
+                break;
+
+            // find next space
+            size_t end = start;
+            while (end < text.size() && text[end] != ' ')
+                ++end;
+
+            tokens.emplace_back(text.data() + start, end - start);
+            pos = end;
+        }
+    }
+
+    void join(const std::string_view * start, size_t n, std::string & ngram) const
+    {
+        size_t total = n - 1; // spaces
+        for (size_t i = 0; i < n; ++i)
+            total += start[i].size();
+        ngram.resize(total);
+
+        size_t pos = 0;
+        for (size_t i = 0; i < n; ++i)
+        {
+            if (i)
+                ngram[pos++] = ' ';
+            memcpy(ngram.data() + pos, start[i].data(), start[i].size());
+            pos += start[i].size();
+        }
+    }
+};
+
+
+template <Tokenizer Tok>
+class NaiveBayesClassifier
+{
+private:
+    /// N-gram size
+    const UInt32 n;
+
+    /// Laplace smoothing parameter
+    const double alpha;
+
+    /// Start / end tokens to pad the input string - taken from Tok at compile time
+    const String start_token;
+    const String end_token;
+
+    Tok tokenizer;
+
+    NGramMap ngram_counts;
+    ClassCountMap class_totals;
+
+    /// Precomputed prior ratios for each class
+    ProbabilityMap class_priors;
+
+    /// Vocabulary size is the number of distinct tokens in the model across all classes
+    size_t vocabulary_size = 0;
+
+    /// Arena to own all the key strings
+    Arena pool;
+
+    inline StringRef allocateString(const String & s)
+    {
+        char * pos = pool.alloc(s.size());
+        memcpy(pos, s.data(), s.size());
+        return StringRef(pos, s.size());
+    }
+
+public:
+    NaiveBayesClassifier() = delete;
+    NaiveBayesClassifier(const NaiveBayesClassifier &) = delete;
+    NaiveBayesClassifier & operator=(const NaiveBayesClassifier &) = delete;
+
+    NaiveBayesClassifier(NaiveBayesClassifier &&) noexcept = default;
+    NaiveBayesClassifier & operator=(NaiveBayesClassifier &&) noexcept = default;
+
+    ~NaiveBayesClassifier() = default;
+
+    /// The model at model_path is expected to be serialized lines of: <class_id> <ngram> <count>
+    NaiveBayesClassifier(
+        const String & model_name,
+        const String & model_path,
+        ProbabilityMap && priors,
+        const UInt32 given_n,
+        const double given_alpha,
+        Tok given_tokenizer = {})
+        : n(given_n)
+        , alpha(given_alpha)
+        , start_token(Tok::start_token)
+        , end_token(Tok::end_token)
+        , tokenizer(std::move(given_tokenizer))
+    {
+        if (!std::filesystem::exists(model_path))
+        {
+            throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "File {} does not exist for model {}", model_path, model_name);
+        }
+
+        DB::ReadBufferFromFile in(model_path);
+
+        while (!in.eof())
```

Comment:
IMO, there should be a test for a malformed model binary, for example, to ensure that it does not crash the server. 
It may be in the form of an integration test (see`tests/integration`).